test/langtools/tools/javac/processing/model/util/elements/TestGetEnumConstantBody.java

```diff
@@ -0,0 +1,258 @@
+            ; // Expected
+        }
+    }
+
+    private void testEnumConstant(VariableElement field,
+                                  TypeElement enclosingClass) {
+        String expectedBinaryName = computeExpectedBinaryName(field);
+        boolean expectEnumConstantBody = expectedBinaryName != null;
+
+        System.out.println("\tTesting enum constant " + field + " expected " + expectEnumConstantBody);
+        if (vacuousElements.getEnumConstantBody(field) != null) {
+            messager.printError("Unexpected vacuous body returned", field);
+        }
+
+        TypeElement enumConstantBody = elements.getEnumConstantBody(field);
+
+        if (Objects.nonNull(enumConstantBody) != expectEnumConstantBody) {
+            messager.printError("Unexpected body value", field);
+        }
+
+        if (enumConstantBody != null) {
+            testEnumConstantBody(enumConstantBody, expectedBinaryName, enclosingClass);
+        }
+
+        System.out.println("\t constant body " + enumConstantBody);
+    }
+
+    /*
+     * From JLS 8.9.1:
+     *
+     * "The optional class body of an enum constant implicitly
+     * declares an anonymous class (15.9.5) that (i) is a direct
+     * subclass of the immediately enclosing enum class (8.1.4), and
+     * (ii) is final (8.1.1.2). The class body is governed by the
+     * usual rules of anonymous classes; in particular it cannot
+     * contain any constructors. Instance methods declared in these
+     * class bodies may be invoked outside the enclosing enum class
+     * only if they override accessible methods in the enclosing enum
+     * class (8.4.8)."
+     */
+    private void testEnumConstantBody(TypeElement enumConstBody, String expectedBinaryName, TypeElement enumClass) {
+        if (enumConstBody.getNestingKind() != NestingKind.ANONYMOUS) {
+            messager.printError("Class body not an anonymous class", enumConstBody);
+        }
+
+        // Get the TypeElement for the direct superclass.
+        TypeElement superClass =
+            (TypeElement)(((DeclaredType)enumConstBody.getSuperclass()).asElement());
+
+        if (!superClass.equals(enumClass)) {
+            messager.printError("Class body is not a direct subclass of the enum", enumConstBody);
+        }
+
+        if (!enumConstBody.getModifiers().contains(Modifier.FINAL)) {
+            messager.printError("Modifier final missing on class body", enumConstBody);
+        }
+
+        if (!elements.getBinaryName(enumConstBody).contentEquals(expectedBinaryName)) {
+            messager.printError("Unexpected binary name, expected: " + expectedBinaryName +
+                                                       ", but was: " + elements.getBinaryName(enumConstBody), enumConstBody);
+        }
+
+        return;
+    }
+
+
+    @interface ExpectedBinaryName {
+        String value();
+    }
+
+    // Nested classes hosting a variety of different kinds of fields.
+
+    private static enum Body {
```

Comment:
I think it makes sense to expand testing, to make sure there are no surprises with constants that override or implement methods, as those are the important cases as per JLS 8.9.1, which you quote.